hazelcast/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -0,0 +1,163 @@
+    public static final QueryDataType SMALLINT = new QueryDataType(ShortConverter.INSTANCE, PRECISION_SMALLINT);
+    public static final QueryDataType INT = new QueryDataType(IntegerConverter.INSTANCE, PRECISION_INT);
+    public static final QueryDataType BIGINT = new QueryDataType(LongConverter.INSTANCE, PRECISION_BIGINT);
+    public static final QueryDataType DECIMAL = new QueryDataType(BigDecimalConverter.INSTANCE);
+    public static final QueryDataType DECIMAL_BIG_INTEGER = new QueryDataType(BigIntegerConverter.INSTANCE);
+    public static final QueryDataType REAL = new QueryDataType(FloatConverter.INSTANCE);
+    public static final QueryDataType DOUBLE = new QueryDataType(DoubleConverter.INSTANCE);
+
+    public static final QueryDataType TIME = new QueryDataType(LocalTimeConverter.INSTANCE);
+    public static final QueryDataType DATE = new QueryDataType(LocalDateConverter.INSTANCE);
+    public static final QueryDataType TIMESTAMP = new QueryDataType(LocalDateTimeConverter.INSTANCE);
+    public static final QueryDataType TIMESTAMP_WITH_TZ_DATE = new QueryDataType(DateConverter.INSTANCE);
+    public static final QueryDataType TIMESTAMP_WITH_TZ_CALENDAR = new QueryDataType(CalendarConverter.INSTANCE);
+    public static final QueryDataType TIMESTAMP_WITH_TZ_INSTANT = new QueryDataType(InstantConverter.INSTANCE);
+    public static final QueryDataType TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME = new QueryDataType(OffsetDateTimeConverter.INSTANCE);
+    public static final QueryDataType TIMESTAMP_WITH_TZ_ZONED_DATE_TIME = new QueryDataType(ZonedDateTimeConverter.INSTANCE);
+
+    public static final QueryDataType OBJECT = new QueryDataType(ObjectConverter.INSTANCE);
+
+    /** Converter. */
+    private Converter converter;
+
+    /** Precision. */
+    private int precision;
+
+    public QueryDataType() {
+        // No-op.
+    }
+
+    QueryDataType(Converter converter) {
+        this(converter, PRECISION_UNLIMITED);
+    }
+
+    QueryDataType(Converter converter, int precision) {
+        this.converter = converter;
+        this.precision = precision;
+    }
+
+    public QueryDataTypeFamily getTypeFamily() {
+        return converter.getTypeFamily();
+    }
+
+    public Converter getConverter() {
+        return converter;
+    }
+
+    public int getPrecision() {
+        return precision;
+    }
+
+    @Override
+    public int getFactoryId() {
+        return QuerySerializationHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return QuerySerializationHook.QUERY_DATA_TYPE;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeInt(converter.getId());
+        out.writeInt(precision);
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        converter = Converters.getConverter(in.readInt());
+        precision = in.readInt();
+    }
```

Comment:
(De)serialization is not covered in tests.

Reply:
I forgot to add it here but then added these tests in the subsequent PR, which will introduce a number of other serializable classes. Let's look at it in the next PR to save on re-review time? Alternatively, I may backport it.

Reply:
Just don't forget about it ;)